packages/mobx-state-tree/src/core/action.ts

```diff
@@ -126,41 +135,61 @@ export function decorate<T extends Function>(middleware: IMiddlewareHandler, fn:
  *
  * @export
  * @template T
- * @param {IMiddlewareHandler} middleware
+ * @param {IMiddlewareHandler} handler
  * @param Function} fn
  * @returns the original function
  */
-export function decorate<T extends Function>(middleware: IMiddlewareHandler, fn: any) {
+export function decorate<T extends Function>(handler: IMiddlewareHandler, fn: any) {
+    const middleware: IMiddleware = { handler, includeHooks: true }
     if (fn.$mst_middleware) fn.$mst_middleware.push(middleware)
     else fn.$mst_middleware = [middleware]
     return fn
 }
 
-function collectMiddlewareHandlers(
+function collectMiddlewares(
     node: ObjectNode,
     baseCall: IMiddlewareEvent,
     fn: Function
-): IMiddlewareHandler[] {
-    let handlers: IMiddlewareHandler[] = (fn as any).$mst_middleware || EMPTY_ARRAY
+): IMiddleware[] {
+    let middlewares: IMiddleware[] = (fn as any).$mst_middleware || EMPTY_ARRAY
     let n: ObjectNode | null = node
     // Find all middlewares. Optimization: cache this?
     while (n) {
-        if (n.middlewares) handlers = handlers.concat(n.middlewares)
+        if (n.middlewares) middlewares = middlewares.concat(n.middlewares)
         n = n.parent
     }
-    return handlers
+    return middlewares
 }
 
 function runMiddleWares(node: ObjectNode, baseCall: IMiddlewareEvent, originalFn: Function): any {
-    const handlers = collectMiddlewareHandlers(node, baseCall, originalFn)
+    const middlewares = collectMiddlewares(node, baseCall, originalFn)
     // Short circuit
-    if (!handlers.length) return mobxAction(originalFn).apply(null, baseCall.args)
+    if (!middlewares.length) return mobxAction(originalFn).apply(null, baseCall.args)
     let index = 0
 
     function runNextMiddleware(call: IMiddlewareEvent): any {
-        const handler = handlers[index++]
-        if (handler) return handler(call, runNextMiddleware)
-        else return mobxAction(originalFn).apply(null, baseCall.args)
+        const middleware = middlewares[index++]
+        const handler = middleware && middleware.handler
+        const invokeHandler = () => {
+            const next = handler(call, runNextMiddleware)
+            if (!next && index < middlewares.length && process.env.NODE_ENV !== "production") {
+                const node = getStateTreeNode(call.tree)
+                fail(
```

Comment:
I'm not sure about this check, there are cases where it is fine that `next()` isn't called, for example if you have middleware that prevents modifications, or only applies it to a copy, etc. I had a few cases before. I am aware it is pretty tricky if you forget it in your middleware, but I think that is an inherent risk about writing your own middleware and hopefully docs help preventing mistakes there :)

Reply:
fair point about aborting the middleware queue.
what about returning false if it the abort was intended and only throw on undefined?

Reply:
e.g. an `abort()` next to `next()`